### ISSUE TITLE
Add a derivative of the mixture fraction or progress variable over time

### ISSUE THREAD

**Issue by czc-zju** [AMReX-Combustion/PeleLMeX#104]
![image](https://user-images.githubusercontent.com/81547457/176627411-3e8f0d5a-8e63-4020-aa77-9280053fb493.png)
How to add a derivative of the mixture fraction or progress variable over time

**Comment by esclapez** [comment]
I see at least two ways to get those: either combining the time derivatives of mass fractions, already available in the advance function or computing mixture fraction and progress variable on both the old and new states, and getting the time derivative, in the `WritePlotFile` function for instance.

**Comment by czc-zju** [comment]
![image](https://user-images.githubusercontent.com/81547457/176990918-eabee810-d46a-4d5c-abcc-55965a860621.png)
I want to get the flame displacement velocity according to this formula. According to the reaction rate and diffusion coefficient, sd is obtained by yt post-processing, but the result is not good compared to that of S3D. The main reason is that after the diffusion term is derived twice, the result is not accurate. S3D directly outputs Dc/Dt inside the program, the result is very good, so I also want to use PeleLMex to output this directly.
![sd](https://user-images.githubusercontent.com/81547457/176992749-891fc90f-d05c-45e7-bd77-8f7f71f75c66.png)

**Comment by czc-zju** [comment]
@[esclapez](https://github.com/esclapez) My purpose is to get the flame displacement velocity, I need to output the **_Fm_** species diffusion fluxes or the divergence of the **_Fm_** , the calculation equation of **_Fm_** is as follows, but in PeleLMeX, how should I output **_Fm_** or **_Vm_**
![image](https://user-images.githubusercontent.com/81547457/177027257-fed68029-780d-4129-9ab5-782354cdbd17.png)
![image](https://user-images.githubusercontent.com/81547457/177027302-f73e3671-bbd7-4d56-ac96-68b18b1fb84b.png)
![image](https://user-images.githubusercontent.com/81547457/177027309-40bc57a6-e4b2-45a5-94e4-b734466ee87d.png)

**Comment by drummerdoc** [comment]
In “advance” you want to compute (Dn + Dnp1)/2 on the last SDC iteration.

**Comment by czc-zju** [comment]
@esclapez @drummerdoc   I've been learning PeleLMeX software for three months now and I'm still trying to learn it. Your contribution is really great and helped me a lot, but I still don't know how to get these parameters ( (**Dn + Dnp1)/2** ) directly. I sincerely hope that you can help me, or you can share some tutorials that can be referenced, thank you very much

**Comment by czc-zju** [comment]
I have tried to output, but the compilation will report an error, my method is still wrong, I sincerely hope that you can help me

**Comment by esclapez** [comment]
@czc-zju getting the displacement speed, and I'm guessing in the future its different components (S_r, S_{d,t}, S_{d,n},...) is a relatively heavy post-processing which could be part of one of [PeleAnalysis](https://github.com/AMReX-Combustion/PeleAnalysis) set of tools (evaluating the terms post-mortem, less consistent with the actual advance of PeleLMeX, but easier) or can be added as an [evaluate derived](https://github.com/AMReX-Combustion/PeleLMeX/blob/development/Source/PeleLMEvaluate.cpp), in which case it can be very close to the actual terms used in PeleLMeX advance.
How familiar are you with the code itself at this point ?

**Comment by czc-zju** [comment]
Thank you for your reply!!!

![s1](https://user-images.githubusercontent.com/81547457/177476716-cf405adf-45a6-4e9c-be9b-f0f56f97437d.png)
It can be seen that **sd =(diffusion term +the source term)/(density and modulus length)**, according to diffusion coefficients of fuel and reaction rate of fuel, I have calculated it in **yt**, but the result of the diffusion term of fuel is not accurate. So I want to output the diffusion term of fuel directly from PeleLMeX, it should be **D(ϕ)** in the following formula: **[[ϕt=F≡A(ϕ)+D(ϕ)+R(ϕ)]].(https://pelelm.readthedocs.io/en/latest/Model.html#misdc-correction-equations)**
I've been studying your commit log these days, it's very helpful to me, but I haven't been able to modify the source code accurately, I'm very sorry. 
Thank you again

**Comment by drummerdoc** [comment]
Yeah, it seems that what you want to compute here really is at the core of what the code computes to advance the solution.  There are lots of intermediate data structures in the code that you can take advantage of, but it really requires a thorough understanding, not only of the algorithm, but of how to manipulate and reward/write AMReX data structures.  We could post a bunch of code of suggestions here, but it might be more productive to set up a meeting to discuss some of the nuances required.  In the best of cases, we can figure something out and not only makes it available to all, but improve the documentation for all as well.  Please feel free to reach out to me at Marc.Day@nrel.gov to see if we can schedule some time at a whiteboard to talk some of this through.

**Comment by czc-zju** [comment]
Thank you very much for your replies.
At present, my ability is not able to make it successfully, but I am very grateful that you will spend time on this, and I am very happy to discuss it.
I think the calculation of **sd** is very important in combustion research and eager to get it.
Below is my process of calculating sd in **yt**.
[sd_python.txt](https://github.com/AMReX-Combustion/PeleLMeX/files/9060876/sd_python.txt)
